File: DesignPatterns/01_GOF23/DP_examples/Lua/03_Behavioral/Iterator/ConcreteAggregate.lua
local IteratorModule = require("Iterator")
local Iterator = IteratorModule.Iterator
local IAggregate = IteratorModule.IAggregate

local defaultCapacity = 4

-- 具体聚合类
local ConcreteAggregate = {}
ConcreteAggregate.__index = ConcreteAggregate
setmetatable(ConcreteAggregate, IAggregate)

function ConcreteAggregate:new(capacity)
    local o = IAggregate:new()
    setmetatable(o, self)
    o.capacity = capacity or defaultCapacity
    if o.capacity < defaultCapacity then
        o.capacity = defaultCapacity
    end
    o.array = {}
    return o
end

function ConcreteAggregate:Count()
    return #self.array
end

function ConcreteAggregate:Add(item)
    table.insert(self.array, item)
end

function ConcreteAggregate:Remove(item)
    for i, v in ipairs(self.array) do
        if v == item then
            table.remove(self.array, i)
            return true
        end
    end
    return false
end

function ConcreteAggregate:Contains(item)
    for _, v in ipairs(self.array) do
        if v == item then
            return true
        end
    end
    return false
end

function ConcreteAggregate:Clear()
    self.array = {}
end

-- 具体迭代器实现
local AggregateIterator = {}
AggregateIterator.__index = AggregateIterator
setmetatable(AggregateIterator, Iterator)

function AggregateIterator:new(aggregate)
    local o = Iterator:new()
    setmetatable(o, self)
    o.aggregate = aggregate
    o.index = 0
    return o
end

function AggregateIterator:First()
    if #self.aggregate.array > 0 then
        self.index = 1
    end
end

function AggregateIterator:Next()
    if self.index <= #self.aggregate.array then
        self.index = self.index + 1
    end
end

function AggregateIterator:IsDone()
    return self.index > #self.aggregate.array
end

function AggregateIterator:CurrentItem()
    if self.index >= 1 and self.index <= #self.aggregate.array then
        return self.aggregate.array[self.index]
    end
    return nil
end

function ConcreteAggregate:GetIterator()
    return AggregateIterator:new(self)
end

return ConcreteAggregate

File: DesignPatterns/01_GOF23/DP_examples/Rust/03_Behavioral/Iterator/concrete_aggregate.rs
use crate::iterator::{IAggregate, Iterable, Iterator};

const DEFAULT_CAPACITY: usize = 4;

// 具体聚合类
pub struct ConcreteAggregate<T> {
    array: Vec<T>,
    capacity: usize,
}

impl<T: PartialEq + Clone> ConcreteAggregate<T> {
    pub fn new(capacity: usize) -> Self {
        let capacity = if capacity < DEFAULT_CAPACITY { DEFAULT_CAPACITY } else { capacity };
        Self {
            array: Vec::with_capacity(capacity),
            capacity,
        }
    }
}

impl<T: PartialEq + Clone> IAggregate<T> for ConcreteAggregate<T> {
    fn count(&self) -> usize {
        self.array.len()
    }

    fn add(&mut self, item: T) {
        self.array.push(item);
    }

    fn remove(&mut self, item: &T) -> bool {
        if let Some(index) = self.array.iter().position(|x| x == item) {
            self.array.remove(index);
            true
        } else {
            false
        }
    }

    fn contains(&self, item: &T) -> bool {
        self.array.contains(item)
    }

    fn clear(&mut self) {
        self.array.clear();
    }
}

impl<T: PartialEq + Clone> Iterable<T> for ConcreteAggregate<T> {
    fn get_iterator(&self) -> Box<dyn Iterator<T> + '_> {
        Box::new(AggregateIterator::new(self))
    }
}

// 具体迭代器实现
pub struct AggregateIterator<'a, T> {
    aggregate: &'a ConcreteAggregate<T>,
    index: isize,
}

impl<'a, T: PartialEq + Clone> AggregateIterator<'a, T> {
    fn new(aggregate: &'a ConcreteAggregate<T>) -> Self {
        Self {
            aggregate,
            index: -1,
        }
    }
}

impl<'a, T: PartialEq + Clone> Iterator<T> for AggregateIterator<'a, T> {
    fn first(&mut self) {
        if self.aggregate.count() > 0 {
            self.index = 0;
        }
    }

    fn next(&mut self) {
        if self.index < self.aggregate.count() as isize {
            self.index += 1;
        }
    }

    fn is_done(&self) -> bool {
        self.index >= self.aggregate.count() as isize
    }

    fn current_item(&self) -> Option<T> {
        if self.index >= 0 && (self.index as usize) < self.aggregate.count() {
            Some(self.aggregate.array[self.index as usize].clone())
        } else {
            None
        }
    }
}
